akka-typed/src/main/scala/akka/typed/javadsl/BehaviorBuilder.scala

```diff
@@ -0,0 +1,141 @@
+    handlers match {
+      case Case(cls, predicate, handler) :: tail ⇒
+        if (cls.isAssignableFrom(msg.getClass) && (predicate.isEmpty || predicate.get.apply(msg))) handler(ctx, msg)
+        else receive[M](ctx, msg, tail)
+      case _ ⇒
+        // emulate scala match error
+        throw new MatchError(msg)
+    }
+
+  /**
+   * Add a new case to the message handling.
+   *
+   * @param `type` type of message to match
+   * @param handler action to apply if the type matches
+   * @tparam M type of message to match
+   * @return a new behavior with the specified handling appended
+   */
+  def message[M <: T](`type`: Class[M], handler: Function2[ActorContext[T], M, Behavior[T]]): BehaviorBuilder[T] =
+    copy[T](messageHandlers = messageHandlers :+ Case[T, T](`type`, None, (i1: ActorContext[T], msg: T) ⇒ handler.apply(i1, msg.asInstanceOf[M])))
+
+  /**
+   * Add a new predicated case to the message handling.
+   *
+   * @param `type` type of message to match
+   * @param test a predicate that will be evaluated on the argument if the type matches
+   * @param handler action to apply if the type matches and the predicate returns true
+   * @tparam M type of message to match
+   * @return a new behavior with the specified handling appended
+   */
+  def message[M <: T](`type`: Class[M], test: Predicate[M], handler: Function2[ActorContext[T], M, Behavior[T]]): BehaviorBuilder[T] =
+    copy[T](messageHandlers = messageHandlers :+ Case[T, T](`type`, Some((t: T) ⇒ test.test(t.asInstanceOf[M])), (i1: ActorContext[T], msg: T) ⇒ handler.apply(i1, msg.asInstanceOf[M])))
+
+  /**
+   * Add a new case to the message handling without compile time type check.
+   * Should normally not be used, but when matching on class with generic type
+   * argument it can be useful, e.g. <code>List.class</code> and <code>(List&lt;String&gt; list) -> {...}</code>
+   *
+   * @param `type` type of message to match
+   * @param handler action to apply when the type matches
+   * @return a new behavior with the specified handling appended
+   */
+  def messageUnchecked[M <: T](`type`: Class[_ <: T], handler: Function2[ActorContext[T], M, Behavior[T]]): BehaviorBuilder[T] =
+    copy[T](messageHandlers = messageHandlers :+ Case[T, T](`type`, None, (i1: ActorContext[T], msg: T) ⇒ handler.apply(i1, msg.asInstanceOf[M])))
+
+  /**
+   * Add a new case to the message handling matching equal messages.
+   *
+   * @param msg the message to compare to
+   * @param handler action to apply when the message matches
+   * @return a new behavior with the specified handling appended
+   */
+  def messageEquals(msg: T, handler: Function[ActorContext[T], Behavior[T]]): BehaviorBuilder[T] =
+    copy[T](messageHandlers = messageHandlers :+ Case[T, T](msg.getClass, Some(_.equals(msg)), (ctx: ActorContext[T], _: T) ⇒ handler.apply(ctx)))
+
+  /**
+   * Add a new case to the signal handling
+   *
+   * @param `type` type of signal to match
```

Comment:
does this look good in javadoc? (try genjavadoc or just rename to clazz or something similar?)